qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,273 @@
+    stored in a single database in qutebrowser's cache directory.
+
+    Top level indices represent categories, each of which is backed by a single
+    table. Child indices represent rows of those tables.
+
+    Class Attributes:
+        COLUMN_WIDTHS: The width percentages of the columns used in the
+                       completion view.
+
+    Attributes:
+        column_widths: The width percentages of the columns used in the
+                        completion view.
+        columns_to_filter: A list of indices of columns to apply the filter to.
+        pattern: Current filter pattern, used for highlighting.
+        _categories: The category tables.
+    """
+
+    def __init__(self, column_widths=(30, 70, 0), columns_to_filter=None,
+                 delete_cur_item=None, parent=None):
+        super().__init__(parent)
+        self.columns_to_filter = columns_to_filter or [0]
+        self.column_widths = column_widths
+        self._categories = []
+        self.srcmodel = self  # TODO: dummy for compat with old API
+        self.pattern = ''
+        self.delete_cur_item = delete_cur_item
+
+    def _cat_from_idx(self, index):
+        """Return the category pointed to by the given index.
+
+        Args:
+            idx: A QModelIndex
+        Returns:
+            A _SqlCompletionCategory if the index points at one, else None
+        """
+        # items hold an index to the parent category in their internalPointer
+        # categories have an empty internalPointer
+        if index.isValid() and not index.internalPointer():
+            return self._categories[index.row()]
+
+    def new_category(self, name, select='*', where=None, sort_by=None,
+                     sort_order=None, limit=None):
+        """Create a new completion category and add it to this model.
+
+        Args:
+            name: Name of category, and the table in the database.
+            select: A custom result column expression for the select statement.
+            where: An optional clause to filter out some rows.
+            sort_by: The name of the field to sort by, or None for no sorting.
+            sort_order: Sorting order, if sort_by is non-None.
+            limit: Maximum row count to return on a query.
+
+        Return: A new CompletionCategory.
+        """
+        cat = _SqlCompletionCategory(name, parent=self, sort_by=sort_by,
+                                     sort_order=sort_order, limit=limit,
+                                     select=select, where=where,
+                                     columns_to_filter=self.columns_to_filter)
+        self._categories.append(cat)
+
+    def data(self, index, role=Qt.DisplayRole):
+        """Return the item data for index.
+
+        Override QAbstractItemModel::data.
+
+        Args:
+            index: The QModelIndex to get item flags for.
+
+        Return: The item data, or None on an invalid index.
```

Comment:
From the Qt docs:

> Note: If you do not have a value to return, return an invalid QVariant instead of returning 0.

Does returning `None` do that? Wouldn't you need to return `QVariant()` instead?